Improve EVP SKEY Implementation

simo5 opened the pull request:
This is a followup to #25908 

With Dmitry we have discussed a few aspects and I wrote some changes I'd like to propose via this separate PR that is based on Dmitry's PR.

The reason to have a separate one is that I am changing some basic things that we may want to discuss before commit to this path. Then we can either integrate this changes on the original PR or supersede it.


The main changes here are:

1. Do away with special casing of "raw keys
2. Implementing full compatibility with unmodified providers
3. Adding fallbacks and additional interface changes to make it possible to have KDF derives use SKEYs both for input and output.

So on (1), the special casing was causing the creation of parallel code paths that would have required exra maintenance, risk difference in semantics between a "real" skey and a mock skey with embedded raw bytes and prevented testing the actual skey machinery with simple tests.

The new code, instead, creates an actual SKEY implementation for the default provider, including SKEYMGMT facility.
Adds fallbacks to import/export keys between providers if neeeded.
Add fallbacks to be able to use SKEY functionality if the cipher provider does not in fact support the new SKEY functionality.

For (2), this whole backward compatibile machinery is demonstrated as working with the test that uses "DES" keys from the legacy provider that I expect will never be changed to use the new interfaces.
While the full SKEY-enabled code is tested with the AES Ciphers that are converted to actually support using SKEYs directly.

And (3) the KDF code has no tests yet but it was a good show-case of a somewhat complex problem to tackle to show how to make backwards compatible code that emits an SKEY directly. (Key generation would be another example, but it doesn't cross providers (from KDF to SKEYMGMT) like the KDF code does).

The patches for the KDF still have a few FIXME along deficiencies in the data available through the exiting KDF interfaces that will probably need to be solved by adding new _ex() functions to create those operations contexts.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

beldmit commented on crypto/evp/evp_enc.c:
```diff
@@ -1837,7 +2024,7 @@ static void *evp_cipher_from_algorithm(const int name_id,
         }
     }
-    if ((fnciphcnt != 0 && fnciphcnt != 3 && fnciphcnt != 4)
+    if ((fnciphcnt != 0 && fnciphcnt != 3 && fnciphcnt != 4 && fnciphcnt != 6)
```
@openssl/committers what should be the logic here? Which callback combinations are mandatory to implement?

beldmit replied:
Fixed in my latest commit, resolving

beldmit commented on crypto/evp/kdf_lib.c:
```diff
@@ -144,6 +145,164 @@ int EVP_KDF_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t keylen,
+
+    if (ctx->meth->set_key != NULL) {
+        if (ctx->meth->prov != key->skeymgmt->prov) {
+            /* TODO: export/import dance */
```
@simo5 can we borrow the export-import code from the cipher implementation?

simo5 replied:
What about a helper function ?

beldmit replied:
Yes, that what I mean

beldmit commented on crypto/evp/s_lib.c:
```diff
@@ -0,0 +1,173 @@
+    if (skeymgmt == NULL) {
+        /* if the specific key_type is unkown, attempt to use the generic
+         * key management */
+        skeymgmt = EVP_SKEYMGMT_fetch(libctx, "GENERIC_SECRET", propquery);
```
Should be a global define

simo5 replied:
You mean "GENERIC_SECRET" ?
Suggestions on where to add it as global ?

simo5 replied:
done

beldmit commented on crypto/evp/skeymgmt_meth.c:
```diff
@@ -0,0 +1,193 @@
+
+void *evp_skeymgmt_import(const EVP_SKEYMGMT *skeymgmt, int selection, const OSSL_PARAM params[])
+{
+    void *provctx = ossl_provider_ctx(EVP_SKEYMGMT_get0_provider(skeymgmt));
```
Still not sure how do you plan to use it

simo5 replied:
Inside the pkcs11-provider code I need the provider's context to perform import operation into the token.

beldmit replied:
And what about export?

beldmit commented on crypto/evp/skeymgmt_meth.c:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+    /*
+     * Try to check that the method is sensible.
```
Outdated comment

beldmit commented on doc/designs/evp_skey.md:
```diff
@@ -0,0 +1,165 @@
+EVP_SKEY *EVP_SKEY_import(OSSL_LIB_CTX *libctx, const char *skeymgmtname,
+                          const char *propquery,
+                          int selection, const OSSL_PARAM *params);
+EVP_SKEY *EVP_SKEY_new_raw(OSSL_LIB_CTX *libctx, const char *key, size_t keylen);
```
To be updated

beldmit commented on doc/man7/provider-cipher.pod:
```diff
@@ -102,11 +108,13 @@ macros in L<openssl-core_dispatch.h(7)>, as follows:
+ OSSL_FUNC_cipher_decrypt_skey_init    OSSL_FUNC_CIPHER_DECRYPT_SKEY_INIT
+ OSSL_FUNC_cipher_update               OSSL_FUNC_CIPHER_UPDATE
+ OSSL_FUNC_cipher_final                OSSL_FUNC_CIPHER_FINAL
+ OSSL_FUNC_cipher_cipher               OSSL_FUNC_CIPHER_CIPHER
```
Nit: Unnecessary whitespace change

beldmit commented on doc/man7/provider-kdf.pod:
```diff
@@ -26,6 +26,7 @@ provider-kdf - The KDF library E<lt>-E<gt> provider functions
  int OSSL_FUNC_kdf_reset(void *kctx);
  int OSSL_FUNC_kdf_derive(void *kctx, unsigned char *key, size_t keylen,
                           const OSSL_PARAM params[]);
+ int OSSL_FUNC_kdf_derive_skey(void *kctx, void *key, const OSSL_PARAM params[]);
```
OSSL_FUNC_kdf_set_skey is missing

beldmit commented on include/crypto/evp.h:
```diff
@@ -246,6 +247,8 @@ struct evp_kdf_st {
     OSSL_FUNC_kdf_get_params_fn *get_params;
     OSSL_FUNC_kdf_get_ctx_params_fn *get_ctx_params;
     OSSL_FUNC_kdf_set_ctx_params_fn *set_ctx_params;
+    OSSL_FUNC_kdf_set_key_fn *set_key;
```
set_skey

beldmit commented on include/crypto/evp.h:
```diff
@@ -751,6 +756,15 @@ struct evp_pkey_st {
+    CRYPTO_RWLOCK *lock;
+
+    void *keydata; /* Alg-specific key data */
+    EVP_SKEYMGMT *skeymgmt; /* Import, export, mana */
```
Nit: manage

beldmit commented on providers/implementations/ciphers/ciphercommon.c:
```diff
@@ -242,6 +244,32 @@ int ossl_cipher_generic_dinit(void *vctx, const unsigned char *key,
+                                   const unsigned char *iv, size_t ivlen,
+                                   const OSSL_PARAM params[])
+{
+    SYM_KEY *aes = skeydata;
```
I'd rename it to smth like skey
Also a blank line after the declaration

beldmit commented on providers/implementations/ciphers/ciphercommon.c:
```diff
@@ -242,6 +244,32 @@ int ossl_cipher_generic_dinit(void *vctx, const unsigned char *key,
+                                   const unsigned char *iv, size_t ivlen,
+                                   const OSSL_PARAM params[])
+{
+    SYM_KEY *aes = skeydata;
```
I'd rename it to smth like skey
also a blank line after declaration

beldmit commented on providers/implementations/skeymgmt/generic.c:
```diff
@@ -0,0 +1,121 @@
+}
+
+/* For any Generic key, we use the "GENERIC_SECRET" algorithm. */
+static const char *rsa_query_operation_name(int operation_id)
```
rsa???

simo5 replied:
copy/paste remnant, will eliminate, there is no query_operation currently in skeymgmt

beldmit commented on test/evp_skey_test.c:
```diff
@@ -0,0 +1,278 @@
+    memset(exported_key, 3, DES_KEY_SIZE);
+
+    /* Do a direct fetch to see it works */
+    des_cbc = EVP_CIPHER_fetch(libctx, "DES-EDE3-CBC", "provider=default");
```
Not sure it tests a legacy provider as it was intended.

beldmit commented:
The test failures are relevant

simo5 commented:
> The test failures are relevant

They may be relevant but there is not enough info to know what is going on.
They clearly pass locally, and in some of the tests, and seem to segfault in some others.

Will try to find out more tomorrow.

simo5 commented:
> > The test failures are relevant
> 
> They may be relevant but there is not enough info to know what is going on. They clearly pass locally, and in some of the tests, and seem to segfault in some others.
> 
> Will try to find out more tomorrow.

Looks like gcc was masking a missing table terminator, while clang did not ... thanks Dmitry for the hint for a reproducer.

I have no idea how to fix the remaining style issues though.

beldmit commented:
When I run check-format to the whole file, I see a lot of complains in the whole block you change so I would waive these failures

beldmit commented:
The only failing style checks are consistent with the surrounding file

kroeckx commented:
CI still has some real issues in it.

simo5 commented:
> CI still has some real issues in it.

Slowly working on them.

simo5 commented:
```
crypto/evp/s_lib.c:118:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
```

34 years later we still cling on C90 ?
Sigh...

kroeckx commented:
> 34 years later we still cling on C90 ? Sigh...

See https://github.com/openssl/technical-policies/issues/33

I guess someone should create a PR.

simo5 commented:
The only remaining failure is a quic test failure in the non-caching test, seem completely unrelated to this commit, but I do not know that code part.

beldmit commented on crypto/evp/s_lib.c:
```diff
@@ -0,0 +1,251 @@
+    skeymgmt = EVP_SKEYMGMT_fetch(libctx, skeymgmtname, propquery);
+    if (skeymgmt == NULL) {
+        /*
+         * if the specific key_type is unkown, attempt to use the generic
```
Nit: unknown

t8m commented on crypto/evp/evp_enc.c:
```diff
@@ -452,6 +466,183 @@ static int evp_cipher_init_internal(EVP_CIPHER_CTX *ctx,
+    /* We have a data managed via key management, using the new callbacks */
+    if (enc) {
+        if (ctx->cipher->einit_skey == NULL) {
+            /* Attempt fallback for providers that do not support SKEYs */
+            struct raw_key_data key = { 0 };
+
+            if (!EVP_SKEY_export(skey, OSSL_SKEYMGMT_SELECT_SECRET_KEY,
+                                 export_key_cb, &key)) {
+                ERR_raise(ERR_LIB_EVP, EVP_R_INITIALIZATION_ERROR);
+                return 0;
+            }
+
+            ret = ctx->cipher->einit(ctx->algctx, key.data, key.length,
+                                     iv, iv_len, params);
+            OPENSSL_free(key.data);
```
I think this export call needs to cache the raw data inside the EVP_SKEY. Otherwise there will be unnecessary performance penalty incurred when using the EVP_SKEY API.

simo5 replied:
We can change the code so that the key data is passed as a const ptr and not as a copied buffer, that way there would be no malloc/free penalty.

In general I expect ciphers to adopt skey natively so this code won't actually end up being used much, so I do not think we should complicate the EVP_SKEY code with caches that then need to be validated for consistency to try to optimize corner cases.

t8m commented:
> So on (1), the special casing was causing the creation of parallel code paths that would have required exra maintenance, risk difference in semantics between a "real" skey and a mock skey with embedded raw bytes and prevented testing the actual skey machinery with simple tests.
> 
> The new code, instead, creates an actual SKEY implementation for the default provider, including SKEYMGMT facility. Adds fallbacks to import/export keys between providers if neeeded. Add fallbacks to be able to use SKEY functionality if the cipher provider does not in fact support the new SKEY functionality.

I think this is a good way forward, however it will need to be augmented with a raw key data cache inside the EVP_SKEY to avoid unnecessary performance penalty if an EVP_SKEY is used with a provided algorithm not supporting the new APIs.

simo5 commented:
> > So on (1), the special casing was causing the creation of parallel code paths that would have required exra maintenance, risk difference in semantics between a "real" skey and a mock skey with embedded raw bytes and prevented testing the actual skey machinery with simple tests.
> > The new code, instead, creates an actual SKEY implementation for the default provider, including SKEYMGMT facility. Adds fallbacks to import/export keys between providers if neeeded. Add fallbacks to be able to use SKEY functionality if the cipher provider does not in fact support the new SKEY functionality.
> 
> I think this is a good way forward, however it will need to be augmented with a raw key data cache inside the EVP_SKEY to avoid unnecessary performance penalty if an EVP_SKEY is used with a provided algorithm not supporting the new APIs.

keydata "cache" already exists for AES and GENERIC_SECERT keys which cover all cases ...

simo5 commented:
@t8m I changed the handling of RAW_BYTES so that only pointers are passed around and copies for key material happen only when really necessary.
I should have done this earlier, copy secret keys around should be avoided as much as possible anyway, not just for performance reasons but also to avoid the need for excessive zeroization.

beldmit commented:
Test failure is relevant

kroeckx commented on doc/man3/EVP_SKEY.pod:
```diff
@@ -0,0 +1,163 @@
+
+B<EVP_SKEY> is a generic structure to hold symmetric keys as opaque objects.
+The keys themselves are often referred to as the "internal key", and are handled by
+providers through L<EVP_KEYMGMT(3)>.
```
EVP_SKEYMGMT

simo5 replied:
This is an issue in @beldmit's PR I am inheriting from

kroeckx commented on doc/man3/EVP_SKEY.pod:
```diff
@@ -0,0 +1,163 @@
+
+The EVP_SKEY_import() function allocates an empty B<EVP_SKEY> structure
+which is used by OpenSSL to store symmetric keys, assigns the
+B<EVP_KEYMGMT> object associated with the key, and initializes the object from
```
EVP_SKEYMGMT

simo5 replied:
This is an issue in @beldmit's PR I am inheriting from

kroeckx commented on doc/man3/EVP_SKEY.pod:
```diff
@@ -0,0 +1,163 @@
+
+ int EVP_SKEY_up_ref(EVP_SKEY *key);
+ void EVP_SKEY_free(EVP_SKEY *key);
+ int EVP_SKEY_is_a(const EVP_SKEY *skey, const char *name);
```
Why don't we just return the type? 

I'm also wondering if this is useful or not

simo5 replied:
I coded it up because I was going to use it, and then it turned out I did not need to use it, I still left it in as we have the same kind of functionality for EVP_PKEY, and may be used later on in interfaces to check that the provide key is of the correct type before calling into providers.

beldmit commented:
@simo5 this failure is relevant
https://github.com/openssl/openssl/actions/runs/12991381032/job/36228752410?pr=26416

beldmit commented:
I tried to combine your commits with mine to make the PR ready for per-commit review and failed

simo5 commented:
> I tried to combine your commits with mine to make the PR ready for per-commit review and failed (in sense that it doesn't simplify the review)

Do you want me to provide fixups against your patches ?

beldmit commented:
> > I tried to combine your commits with mine to make the PR ready for per-commit review and failed (in sense that it doesn't simplify the review)
> 
> Do you want me to provide fixups against your patches ?

Yes if it's clear for you how to do it

simo5 commented:
Rebased on top of Dmitry's branch now that we rebased all commits there.

This PR will be repurposed to test against a custom branch of pkcs11-provider that implements SKEY providers

simo5 commented:
All the work here has been intagrated in the original PR or in #26702 or #26542, so closing this one for now to avoid confusion
